[PRDI-602] Delete Pending Files On Network Failure

brianmichel opened the pull request:
When members use a tool like Little Snitch or pi-hole to block outbound requests to segment.io, the Segment library will persist these analytics to disk into different files. These files can stack up immensely over time. This could mean you are trying to make 1,000s of network calls to upload these analytics that will never succeed. We would prefer to give this CPU time back to members at the trade off of maybe having some fewer events for the time being.

This means that as events fail to upload we will slowly delete the temporary files on disk so that they are not building up over time. This means that we will likely give up some analytics event integrity to give back CPU time and general performance to member machines that are in this state (i.e. people using this kind of blocking software).

brianmichel committed:
Delete Pending Files On Network Failure

When members use a tool like Little Snitch or pi-hole to block outbound requests to segment.io, the Segment library will persist these analytics to disk into different files. These files can stack up immensely over time. This could mean you are trying to make 1,000s of network calls to upload these analytics that will never succeed. We would prefer to give this CPU time back to members at the trade off of maybe having some fewer events for the time being.

brianmichel commented:
This is a slightly more aggressive approach than what Segment is pondering here https://github.com/segmentio/analytics-swift/pull/153/files, but their approach doesn't solve actually preventing 1000s of network request, it seems to simply delay them.

brianmichel commented:
Beep boop, I'm a bot, here's a link to the ticket https://linear.app/the-browser-company/issue/PRDI-602/investigate-segment-tracking-error-scenarios

rgrunberg commented:
This approach sgtm, will let Stephan review the actual code :)

brianmichel commented:
Let me know what you think @diederich, happy to Tuple over this change, but would like to get it merged and into a build sooner rather than later so we could see if there are any knock on effects that haven't been covered in the discussions of this issue on the other threads I've linked.

diederich commented on Sources/Segment/Plugins/SegmentDestination.swift:
```diff
@@ -120,20 +120,25 @@ public class SegmentDestination: DestinationPlugin {
+                            }
                     }
+
+                    analytics.logFlush()
```
Do we wanna call this all the time now? (before it was only in the `default` case it seems?)

brianmichel replied:
Looking at the PR from Segment themselves https://github.com/segmentio/analytics-swift/pull/153/files it appears that they move the `logFlush` out of the switch to happen all of the time since there is no longer a `default` case due to handling the `.failure` case explicitly. 

This would trigger a re-flushing of the pending log files, which I believe will get us into a better state for members stuck in this scenario (i.e this that have blocked segment.io). However, this feels like maybe it could cause some unwanted looping behavior for those members as well (i.e. you fail the call, we delete the file, call logFlush, the log flushing process starts again, and fails quickly due to the API being blocked, and then we're back at the beginning of the loop I think). 

I _could_ move the `logFlush` up into the `failure` handlings since I _believe_ the prior `default` case was really just the error case anyway. Would love some thoughts here.

diederich replied:
Oh - I didn't get there was an upstream fix already. Thought this is gonna be it :-) 
I'm a bit confused now why we're not picking the upstream commits into this, but do our own version of it. 
Also, mind that our connectivityError's are different - `e.g.` ``timedOut` counts as no `connectivityError` upstream, where as we flag it as one. Not sure if on purpose?

The upstream PR does not fix the issue of files piling up though - does it? Reading the PR description it's just about missing the 30s flushs? 

What about pulling the upstream PR in (in case we want it now already) - but independently of it, drop any excess files lying around (with something like the approach in the other comment). And maybe also upstream that?

brianmichel replied:
Happy to pull in the upstream PR once it's landed, but as you've pointed out it's fixing something a little different than this. But we will integrate it once it's in main

diederich commented on Sources/Segment/Plugins/SegmentDestination.swift:
```diff
@@ -120,20 +120,25 @@ public class SegmentDestination: DestinationPlugin {
+                            self?.cleanupUploads()
+                        case let .failure(error):
+                            // Until https://github.com/segmentio/analytics-swift/issues/152 is resolved we should remove failing files since we run the risk of building up a huge list of pending files to upload.
+                            if error.connectivityError {
```
hmhmhm, makes me a little nervous to discard all on network failure. (not sure when flush is triggered, think of short network failures / tunnel / tube / .. :-)) 

Thought about capping the `data` thing we iterate over - e.g. dropping everything > 50 but leaving the rest here untouched? (though we'd need to know the "order")

brianmichel replied:
It looks like flush is triggered quite frequently but here are a couple of key scenarios:
- Foregrounding the app
- Backgrounding the app
- At a somewhat set interval (I believe we have this set to every 10 seconds currently)

A slightly more sophisticated solution here would be tracking failures per file and then maybe deleting a file after X failures? 

I'd venture a guess only a handful of people are in the configuration where they are blocking api.segment.io so I erred on the side of the simplest solution that fixes the problem, while trying to articulate the trade offs that there would likely be some data log around our instrumentation due to this approach for the scenarios that you mention @diederich. It's a bit hard to quantify what the size of data loss would be for non-blocking members, but it's a tradeoff nonetheless. Would love @rgrunberg's opinion here as well since she will be directly impacted.

rgrunberg replied:
Ideally
1) we'd attempt more than once before dropping a file (in case the connectivity problem is temporary)
and/or
2) we'd have some instrumentation of how often there are temporary network issues resulting in event deletion

That being said, I'm not too worried about partial data loss for members who are just experiencing temporary network issues, since our key metrics are fairly robust to that (e.g. if you open 200 tabs in a day and we drop 5% of those logs, you're still going to be counted as an active member). So if these other options seem difficult to implement, I don't think they should block getting a fix in for the main issue here.

diederich replied:
Thanks @rgrunberg!

@brianmichel in the spirit of simple / not changing - maybe, before we get into the `for url in data {`-loop:
```
if data.length > some-max {
   let filesToDrop = data.prefix(data.length - some-max)
   filesToDrop.forEach { storage.remove(file: $0.url) }
   data = data.suffix(some-max)
}
```
and then let the rest of it as is? So we'd just be dropping things in case the pile up, but otherwise keep the retry logic and flushing etc. as is. WDYT?

brianmichel replied:
I like it, let's give it a shot and see how it goes!

brianmichel replied:
Pushed up this solution instead of what I had.

diederich reviewed (commented):
sorry, I missed this ealier.

brianmichel committed:
Return old upload handling code

brianmichel committed:
Creat configurable max log file backlog

We want to be able to proactively clean up log files that are super old. We now have a configuration value that we can dial down to only keep a certain number of fresh analytics logs around.

brianmichel commented on Sources/Segment/Plugins/SegmentDestination.swift:
```diff
@@ -120,20 +133,20 @@ public class SegmentDestination: DestinationPlugin {
                 analytics.log(message: "Processing Batch:\n\(url.lastPathComponent)")
                 
-                let uploadTask = httpClient.startBatchUpload(writeKey: analytics.configuration.values.writeKey, batch: url) { (result) in
+                let uploadTask = httpClient.startBatchUpload(writeKey: analytics.configuration.values.writeKey, batch: url) { [weak self] (result) in
```
I kept the `[weak self]` since it seemed like a good small clean up to have anyway.

brianmichel commented on Sources/Segment/Configuration.swift:
```diff
@@ -30,6 +30,7 @@ public class Configuration {
         var autoAddSegmentDestination: Bool = true
         var apiHost: String = HTTPClient.getDefaultAPIHost()
         var cdnHost: String = HTTPClient.getDefaultCDNHost()
+        var maximumLogFilesOnDisk: Int = 15
```
No idea what the right number is here, happy to change the default...

brianmichel replied:
We're currently configured to flush at every 3 events (we should probably change this, it feels aggressive and overly chatty) and/or every 10 seconds. 

This means with these defaults we'll have a backlog of 45 events spread over 15 files on disk.

brianmichel commented:
Video showing the files on disk being purged at the limit of what we've configured:

https://user-images.githubusercontent.com/63919/191308405-92c8a7ea-5ca2-4093-b5ef-f3a2c4ae4b5e.mp4

brianmichel committed:
Just doing a small clean up

diederich reviewed (approved):
Thaaank you - looks great! 🙏
